be1-go/hub/standard_hub/hub_state/MessageIds.go

```diff
@@ -0,0 +1,32 @@
+package hub_state
+
+import (
+	"golang.org/x/exp/slices"
+)
+
+// MessageIds stores a channel id with its corresponding message ids
+type MessageIds struct {
+	ThreadSafeMap[string, []string]
+}
+
+// Add adds a message id to the slice of message ids of the channel
+func (i *MessageIds) Add(channel string, id string) {
+	i.Lock()
+	defer i.Unlock()
+	messageIds, channelStored := i.table[channel]
+	if !channelStored {
+		i.table[channel] = append(i.table[channel], id)
+		return
+	}
+	alreadyStoredId := slices.Contains(messageIds, id)
```

Comment:
Here we have two conflicting needs, we want messageIds stored sequentially (i.e. order matters) but this access is O(n) and it won't scale well as the number of messages grows.
We could trade memory for speed, and have a secondary map that contains maps as values, and allows us to check in O(1) if an ID is included or not.
Now that you have a `MessageIds` struct, nothing prevents you from optimizing the heck out of it 😁 

I assume this is not the only place where O(n) operations happen, so maybe this is also best addressed in a separate PR, probably when we look at persisting this to disk as well :)

Reply:
I agree that there's a lot of room for optimization here. I opened #1708 to keep track of the issue and discuss it when we start working on persistence.